### ISSUE TITLE
Crash (NullReferenceException) in managed-to-native wrapper on Win64

### ISSUE THREAD

**Issue by joncham** [mono/mono#7596]
## Steps to Reproduce

1. Build the following C++ into a dynamic library
2. Build the sample C# code
3. Run with Mono on Windows 64-bit with debugger sequence points enabled (I'm setting `MONO_DEBUG=gen-seq-points`)

```
// C++
typedef struct
{
	uint16_t a;
	int32_t b; // Boolean comes in as int32_t
	int32_t c; // Boolean comes in as int32_t

} StructToMarshal;

static StructToMarshal local;

extern "C" __declspec(dllexport) void NativeMethod (StructToMarshal options)
{
	local = options;
}
```

```
// C#
using System.Runtime.InteropServices;

class Program
{
    static void Main()
    {
        var o = new StructToMarshal();
        NativeMethod(o);
    }

    [DllImport("DynamicLibrary.dll")]
    static extern void NativeMethod(StructToMarshal options);
    public struct StructToMarshal
    {
        public ushort a;
        public bool b;
        public bool c;
    }
}
```
## Current Behavior

```
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at (wrapper managed-to-native) Program.NativeMethod(Program/StructToMarshal)
  at Program.Main () [0x00009] in <d6a5c184bbdb418bb6ee378ea93cbefb>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at (wrapper managed-to-native) Program.NativeMethod(Program/StructToMarshal)
  at Program.Main () [0x00009] in <d6a5c184bbdb418bb6ee378ea93cbefb>:0
```

## Expected Behavior

No crash

### On which platforms did you notice this

[ ] macOS
[ ] Linux
[ X] Windows

**Version Used**:
```
Mono JIT compiler version 5.11.0 (Visual Studio built mono)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           normal
        SIGSEGV:       normal
        Notification:  Thread + polling
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        GC:            sgen (concurrent by default)
```

**Comment by joncham** [comment]
The actual crash is due to a stack corruption prior to the native call. We overwrite a part of the stack that is expected to point to `ss_trampoline` and when a single step check is performed the pointer is garbage.

**Comment by lateralusX** [comment]
In Windows x64 ABI, structs not fitting into register (or using some odd sizes) will be passed on stack and a reference to the stack address is then passed to the native function, either in a parameter register or using a stack slot. The ABI parts works as expected but what was not working as expected is the sizing of the local variable representing the native struct. That local is currently not flagged as being used in a pinvoke, so the allocation of it will use the managed struct stack size instead of the native. When data is copied into this local variable before doing the pinvoke call, data will be copied using the native struct size overwrite things on the stack, in this case 4 bytes into the local used by the trampoline data.

**Comment by lateralusX** [comment]
Found a similar issue with empty struct returns from pinvoke (0 managed size) when running pinvoke unit tests with sequence points enabled. Will include fix for this in same PR.

**Comment by lateralusX** [comment]
This issue should be fixed by #7625.
